README.md
# MercadoPago SDK


[![Crate version on crates.io](https://img.shields.io/crates/v/mercadopago-sdk-rust)](https://crates.io/crates/mercadopago-sdk-rust)
[![Crate documentation on docs.rs](https://img.shields.io/docsrs/mercadopago-sdk-rust)](https://docs.rs/mercadopago-sdk-rust)
![Crate license](https://img.shields.io/crates/l/mercadopago-sdk-rust)


<!-- cargo-rdme start -->

An open source, strongly-typed SDK for the MercadoPago API.

It will try to hold your hand and reduce the possibility of errors, providing the correct API
surface.

### Note

The library is still under development and its public API is subject to change.

# Installation

Added the following into your Cargo.toml:

```toml
mercadopago_sdk_rust = "0.1"
```

# Usage

The client is built using the
[`MercadoPagoSDKBuilder::with_token`](https://docs.rs/mercadopago-sdk-rust/latest/mercadopago_sdk_rust/struct.MercadoPagoSDKBuilder.html) `with_token`
method.

```rust
use mercadopago_sdk_rust::{MercadoPagoSDK, MercadoPagoSDKBuilder};

let mp_sdk: MercadoPagoSDK = MercadoPagoSDKBuilder::with_token("MP_ACCESS_TOKEN");

```

Once the token is inserted, you can call methods on [`crate::MercadoPagoSDK`]



# Creating a CheckoutPro Preference
```rust
use mercadopago_sdk_rust::common_types::{CheckoutProPayer, Item};
use mercadopago_sdk_rust::payments::requests::DocumentType;
use mercadopago_sdk_rust::preferences::requests::CheckoutProPreferences;
use mercadopago_sdk_rust::MercadoPagoSDKBuilder;

#[tokio::main]
async fn async_main() {
    let mp_sdk = MercadoPagoSDKBuilder::with_token("MP_ACCESS_TOKEN");

    let sample_item =
        Item::minimal_item("Sample item".to_string(), "".to_string(), 15.00, 1).unwrap();

    let preferences = CheckoutProPreferences::new()
        .set_items(vec![sample_item])
        .set_payer(CheckoutProPayer::minimal_payer(
            "fulano@beltrano.com.br".to_string(),
            DocumentType::CPF,
            41810524485,
        ));

    mp_sdk
        .create_preferences_checkout_pro(preferences)
        .expect("Failed to validate checkout preference. Something is wrong.")
        .execute()
        .await
        .unwrap();
}
```

# Other Examples

Check out the `tests` folder inside our repository to check for more examples.

# License
Project is licensed under the permissive MIT license.

<!-- cargo-rdme end -->


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v0.1.1 (2022-05-15)

### Chore

 - <csr-id-4ea57c391e614fb967e0d2ed623f65c803757b47/> changes to .gitignore

 - <csr-id-0f1a72b2e0fe3d46a9e8aec9829296dc0585b99a/> added README.toml, added repo to cargo.toml, added docs


### Documentation

 - <csr-id-612406b26b46bcbbd2b170e036a377304c4716f7/> changes to readme

 - <csr-id-0ed5307a54a5d4cc9be50b050b600c8c26e55411/> added description and license to cargo.toml

 - <csr-id-dd7f8f735fc6f337b56997ac2fe3417ba9dc55e1/> added CHANGELOG


### New Features

 - <csr-id-7eb273ac148faf37885ed57ef208144fe92b1122/> And there be light. Initial commit.


### Commit Statistics

<csr-read-only-do-not-edit/>

 - 9 commits contributed to the release.
 - 6 commits where understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' where seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - changes to readme ([`612406b`](https://github.comgit//saskenuba/mercadopago-sdk-rust/commit/612406b26b46bcbbd2b170e036a377304c4716f7))
    - docs, feat: create_card_token is now working correctly ([`c544b79`](https://github.comgit//saskenuba/mercadopago-sdk-rust/commit/c544b79cfcdb1e9d6f6da11752aeacba043d2dd7))
    - added description and license to cargo.toml ([`0ed5307`](https://github.comgit//saskenuba/mercadopago-sdk-rust/commit/0ed5307a54a5d4cc9be50b050b600c8c26e55411))
    - added CHANGELOG ([`dd7f8f7`](https://github.comgit//saskenuba/mercadopago-sdk-rust/commit/dd7f8f735fc6f337b56997ac2fe3417ba9dc55e1))
    - Release mercadopago-sdk-rust v0.1.0 ([`0d8f632`](https://github.comgit//saskenuba/mercadopago-sdk-rust/commit/0d8f632adadf7794772317755c3b429729ad489b))
    - changes to .gitignore ([`4ea57c3`](https://github.comgit//saskenuba/mercadopago-sdk-rust/commit/4ea57c391e614fb967e0d2ed623f65c803757b47))
    - added README.toml, added repo to cargo.toml, added docs ([`0f1a72b`](https://github.comgit//saskenuba/mercadopago-sdk-rust/commit/0f1a72b2e0fe3d46a9e8aec9829296dc0585b99a))
    - And there be light. Initial commit. ([`7eb273a`](https://github.comgit//saskenuba/mercadopago-sdk-rust/commit/7eb273ac148faf37885ed57ef208144fe92b1122))
    - Initial commit ([`adbc7c4`](https://github.comgit//saskenuba/mercadopago-sdk-rust/commit/adbc7c455eb97bd96ac4c637ec0dcaccd7d23f97))
</details>

## v0.1.0 (2022-05-15)

<csr-id-4ea57c391e614fb967e0d2ed623f65c803757b47/>
<csr-id-0f1a72b2e0fe3d46a9e8aec9829296dc0585b99a/>

### Chore

 - <csr-id-4ea57c391e614fb967e0d2ed623f65c803757b47/> changes to .gitignore

 - <csr-id-0f1a72b2e0fe3d46a9e8aec9829296dc0585b99a/> added README.toml, added repo to cargo.toml, added docs


### New Features

 - <csr-id-7eb273ac148faf37885ed57ef208144fe92b1122/> And there be light. Initial commit.



src/lib.rs
//! An open source, strongly-typed SDK for the MercadoPago API.
//!
//! It will try to hold your hand and reduce the possibility of errors, providing the correct API
//! surface.
//!
//! ### Note
//!
//! The library is still under development and its public API is subject to change.
//!
//! # Installation
//!
//! Added the following into your Cargo.toml:
//!
//! ```toml
//! mercadopago_sdk_rust = "0.1"
//! ```
//!
//! # Usage
//!
//! The client is built using the
//! [`MercadoPagoSDKBuilder::with_token`](crate::MercadoPagoSDKBuilder) `with_token`
//! method.
//!
//! ```rust
//! # fn main() {
//! use mercadopago_sdk_rust::{MercadoPagoSDK, MercadoPagoSDKBuilder};
//!
//! let mp_sdk: MercadoPagoSDK = MercadoPagoSDKBuilder::with_token("MP_ACCESS_TOKEN");
//!
//! # }
//! ```
//!
//! Once the token is inserted, you can call methods on [`crate::MercadoPagoSDK`]
//!
//!
//!
//! # Creating a CheckoutPro Preference
//! ```no_run
//! use mercadopago_sdk_rust::common_types::{CheckoutProPayer, Item};
//! use mercadopago_sdk_rust::payments::requests::DocumentType;
//! use mercadopago_sdk_rust::preferences::requests::CheckoutProPreferences;
//! use mercadopago_sdk_rust::MercadoPagoSDKBuilder;
//!
//! #[tokio::main]
//! async fn async_main() {
//!     let mp_sdk = MercadoPagoSDKBuilder::with_token("MP_ACCESS_TOKEN");
//!
//!     let sample_item =
//!         Item::minimal_item("Sample item".to_string(), "".to_string(), 15.00, 1).unwrap();
//!
//!     let preferences = CheckoutProPreferences::new()
//!         .set_items(vec![sample_item])
//!         .set_payer(CheckoutProPayer::minimal_payer(
//!             "fulano@beltrano.com.br".to_string(),
//!             DocumentType::CPF,
//!             41810524485,
//!         ));
//!
//!     mp_sdk
//!         .create_preferences_checkout_pro(preferences)
//!         .expect("Failed to validate checkout preference. Something is wrong.")
//!         .execute()
//!         .await
//!         .unwrap();
//! }
//! ```
//!
//! # Other Examples
//!
//! Check out the `tests` folder inside our repository to check for more examples.
//!
//! # License
//! Project is licensed under the permissive MIT license.

pub mod card_tokens;
pub mod common_types;
pub mod errors;
pub mod helpers;
pub mod payments;
pub mod preferences;
pub mod webhooks;

use std::marker::PhantomData;

use futures::future::err;
use futures::TryFutureExt;
use oauth2::basic::BasicClient;
use oauth2::reqwest::async_http_client;
use oauth2::{
    AccessToken, AuthType, AuthUrl, ClientId, ClientSecret, Scope, TokenResponse, TokenUrl,
};
use reqwest::{Client, Method, RequestBuilder};
use serde::de::DeserializeOwned;

use crate::card_tokens::requests::CardTokenOptions;
use crate::card_tokens::responses::CardTokenResponse;
use crate::errors::{ApiError, SDKError};
use crate::payments::requests::CreatePaymentPayload;
use crate::preferences::requests::CheckoutProPreferences;
use crate::preferences::responses::CheckoutProPreferencesResponse;

const API_BASE_URL: &str = "https://api.mercadopago.com";

///
#[derive(Debug)]
pub struct MercadoPagoSDKBuilder {}

impl MercadoPagoSDKBuilder {
    async fn authorize<T: ToString>(
        client_id: T,
        client_secret: T,
    ) -> Result<MercadoPagoSDK, SDKError> {
        let client = BasicClient::new(
            ClientId::new(client_id.to_string()),
            Some(ClientSecret::new(client_secret.to_string())),
            AuthUrl::new("https://auth.mercadopago.com/authorization".to_string()).unwrap(),
            Some(TokenUrl::new("https://api.mercadopago.com/oauth/token".to_string()).unwrap()),
        )
        .set_auth_type(AuthType::BasicAuth);

        let token_response = client
            .exchange_client_credentials()
            .add_scope(Scope::new("offline_access".to_string()))
            .request_async(async_http_client)
            .await
            .unwrap();

        Ok(MercadoPagoSDK {
            http_client: Default::default(),
            access_token: token_response.access_token().clone(),
        })
    }

    /// Creates an [`MercadoPagoSDK`] ready to request the API.
    pub fn with_token<T: ToString>(client_access_token: T) -> MercadoPagoSDK {
        MercadoPagoSDK {
            http_client: Default::default(),
            access_token: AccessToken::new(client_access_token.to_string()),
        }
    }
}

#[derive(Debug)]
pub struct MercadoPagoSDK {
    pub(crate) http_client: Client,
    pub(crate) access_token: AccessToken,
}

#[derive(Debug)]
pub struct SDKRequest<'a, RP> {
    http_client: &'a Client,
    access_token: &'a AccessToken,
    request: RequestBuilder,
    response_type: PhantomData<RP>,
}

impl<'a, RP> SDKRequest<'a, RP> {
    /// Injects bearer token, and return response
    pub async fn execute(self) -> Result<RP, SDKError>
    where
        RP: DeserializeOwned,
    {
        let request = self
            .request
            .bearer_auth(self.access_token.secret())
            .build()
            .unwrap();
        let response = self
            .http_client
            .execute(request)
            .and_then(|c| c.text())
            .await?;
        eprintln!("response = {}", response);

        // matches errors due to wrong payloads etc
        let error_jd = serde_json::from_str::<ApiError>(&*response);
        if let Ok(err) = error_jd {
            eprintln!("err = {:#?}", err);
            return Err(SDKError::GenericError);
        }

        let jd = &mut serde_json::Deserializer::from_str(&*response);
        let res: Result<RP, _> = serde_path_to_error::deserialize(jd);

        match res {
            Ok(deserialized_resp) => Ok(deserialized_resp),
            Err(wow) => {
                println!("{:?}", wow.path());
                eprintln!("Error = {:#?}", wow);
                Err(SDKError::GenericError)
            }
        }
    }
}

impl MercadoPagoSDK {
    pub fn create_preferences_checkout_pro(
        &self,
        opts: CheckoutProPreferences,
    ) -> Result<SDKRequest<CheckoutProPreferencesResponse>, SDKError> {
        if opts.validate() {}

        let request = self
            .http_client
            .request(
                Method::POST,
                API_BASE_URL.to_string() + "/checkout/preferences",
            )
            .json(&opts);

        Ok(SDKRequest {
            http_client: &self.http_client,
            access_token: &self.access_token,
            request,
            response_type: PhantomData::<_>,
        })
    }

    /// Used to create and save a credit/debit card token, instead of transacting raw sensitive
    /// data, such as card number.
    ///
    /// Create a token before issuing payments with cards.
    pub fn create_card_token(
        &self,
        opts: CardTokenOptions,
    ) -> Result<SDKRequest<CardTokenResponse>, SDKError> {
        let url = format!(
            "{}/v1/card_tokens?public_key={}",
            API_BASE_URL,
            opts.public_key.as_deref().unwrap_or("")
        );

        let request = self.http_client.request(Method::POST, url).json(&opts);

        Ok(SDKRequest {
            http_client: &self.http_client,
            access_token: &self.access_token,
            request,
            response_type: PhantomData::<_>,
        })
    }

    pub fn create_payment(
        &self,
        opts: CreatePaymentPayload,
    ) -> Result<SDKRequest<CheckoutProPreferencesResponse>, SDKError> {
        let request = self
            .http_client
            .request(Method::POST, "/payments")
            .json(&opts);

        Ok(SDKRequest {
            http_client: &self.http_client,
            access_token: &self.access_token,
            request,
            response_type: PhantomData::<CheckoutProPreferencesResponse>,
        })
    }
}
